torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp

```diff
@@ -0,0 +1,637 @@
+  //    t0
+  //    |
+  //   cache
+  //   /  \
+  //  t1  t2
+  // if groups = {{t1, t2}, {t0}}, then removing {t0, cache} from the DAG will
+  // make it disconnected.
+  std::unordered_set<TensorView*> group2_and_cached_inputs(
+      grouped_inputs_outputs[1].begin(), grouped_inputs_outputs[1].end());
+  for (auto tv : grouped_inputs_outputs[1]) {
+    if (tv->isFusionInput()) {
+      auto existing_cache = ir_utils::consumerTvsOf(tv)[0];
+      if (ir_utils::consumerTvsOf(existing_cache).size() > 1) {
+        auto new_cache = tv->cacheAfter();
+        new_cache->setMemoryType(MemoryType::Shared);
+        group2_and_cached_inputs.emplace(new_cache);
+      } else {
+        existing_cache->setMemoryType(MemoryType::Shared);
+        group2_and_cached_inputs.emplace(existing_cache);
+      }
+    }
+  }
+  // set cached outputs of group 2 to shared memory
+  for (auto pair : cached_outputs) {
+    auto cached_output = pair.first;
+    auto output = pair.second;
+    if (group2_and_cached_inputs.count(output) > 0) {
+      cached_output->setMemoryType(MemoryType::Shared);
+    }
+  }
+
+  TensorView* reference1 =
+      domain_map.findReferenceFor(grouped_inputs_outputs[0]);
+  TensorView* reference2 =
+      domain_map.findReferenceFor(grouped_inputs_outputs[1]);
+
+  TORCH_INTERNAL_ASSERT(
+      reference1 != nullptr,
+      "Could not find a fully broadcasted tensor to reference schedule on the first group.");
+
+  TORCH_INTERNAL_ASSERT(
+      reference2 != nullptr,
+      "Could not find a fully broadcasted tensor to reference schedule on the second group.");
+
+  auto inner_most_id1 = scheduler_utils::innerMostRootDim(reference1);
+  auto inner_most_id2 = scheduler_utils::innerMostRootDim(reference2);
+
+  auto inner_most_pos1_in_ref1 =
+      domain_map.getPosMappedTo(reference1, inner_most_id1);
+  auto inner_most_pos2_in_ref1 =
+      domain_map.getPosMappedTo(reference1, inner_most_id2);
+
+  // make tile
+  // [..., I1, .., I2, ...]
+  reference1->split(inner_most_pos1_in_ref1, params.tile_size1);
+  reference1->reorder({{inner_most_pos1_in_ref1 + 1, -1}});
+  reference1->split(inner_most_pos2_in_ref1, params.tile_size2);
+  reference1->reorder({{inner_most_pos2_in_ref1 + 1, -1}});
+  // [..., I1/tile1, .., I2/tile2, ..., tile1, tile2]
+
+  // Merge remaining dimensions
+  int lhs_i = -1;
+  for (int i = (int)reference1->nDims() - 2; i > 0; i--) {
+    auto axis_i = i - 1;
+    if (reference1->axis(axis_i)->isBroadcast() ||
+        reference1->axis(axis_i)->isReduction()) {
+      continue;
+    }
+    if (lhs_i == -1) {
+      lhs_i = axis_i;
+    } else {
+      reference1->merge(axis_i, lhs_i);
+      lhs_i = axis_i;
+    }
+  }
+  reference1->split(0, 1);
+  // [merged_dim, 1, unmerged dims..., tile1, tile2]
+
+  // parallelize non-tile dimensions
+  reference1->axis(1)->parallelize(ParallelType::Unswitch);
+  reference1->axis(0)->parallelize(ParallelType::BIDx);
+  // [BIDx, Unswitch, unmerged dims..., tile1, tile2]
```

Comment:
yeah, a little worried about unmerged dims if they weren't resolved for this reference. Maybe we're picking a reference that must have them resolved, but that seems like a constraint we don't really need.